topdown/aggregates.go

```diff
@@ -5,88 +5,83 @@
-			return errors.Wrapf(err, "aggregate")
-		}
-
-		y, err := f(x)
-		if err != nil {
-			switch err.(type) {
-			case empty:
-				return nil
+func builtinSum(a ast.Value) (ast.Value, error) {
+	switch a := a.(type) {
+	case ast.Array:
+		sum := big.NewFloat(0)
+		for _, x := range a {
+			n, ok := x.Value.(ast.Number)
+			if !ok {
+				return nil, builtins.NewOperandElementErr(1, a, x.Value, ast.NumberTypeName)
 			}
-			return err
+			sum = new(big.Float).Add(sum, builtins.NumberToFloat(n))
 		}
-
-		undo, err := evalEqUnify(t, y, dst, nil, iter)
-		t.Unbind(undo)
-		return err
-	}
-}
-
-func reduceSum(x interface{}) (ast.Value, error) {
-	if s, ok := x.([]interface{}); ok {
+		return builtins.FloatToNumber(sum), nil
+	case *ast.Set:
 		sum := big.NewFloat(0)
-		for _, x := range s {
-			n, ok := x.(json.Number)
+		for _, x := range *a {
+			n, ok := x.Value.(ast.Number)
 			if !ok {
-				return nil, fmt.Errorf("sum: input elements must be numbers")
+				return nil, builtins.NewOperandElementErr(1, a, x.Value, ast.NumberTypeName)
 			}
-			sum = new(big.Float).Add(sum, jsonNumberToFloat(n))
+			sum = new(big.Float).Add(sum, builtins.NumberToFloat(n))
 		}
-		return floatToASTNumber(sum), nil
+		return builtins.FloatToNumber(sum), nil
 	}
-	return nil, fmt.Errorf("sum: source must be array")
+	return nil, builtins.NewOperandTypeErr(1, a, ast.SetTypeName, ast.ArrayTypeName)
 }
 
-func reduceCount(x interface{}) (ast.Value, error) {
-	switch x := x.(type) {
-	case []interface{}:
-		return ast.IntNumberTerm(len(x)).Value, nil
-	case map[string]interface{}:
-		return ast.IntNumberTerm(len(x)).Value, nil
-	case string:
-		return ast.IntNumberTerm(len(x)).Value, nil
-	default:
-		return nil, fmt.Errorf("count: source must be array, object, or string")
-	}
-}
-
-func reduceMax(x interface{}) (ast.Value, error) {
-	switch x := x.(type) {
-	case []interface{}:
-		if len(x) == 0 {
-			return nil, empty{}
+func builtinMax(a ast.Value) (ast.Value, error) {
+	switch a := a.(type) {
+	case ast.Array:
+		if len(a) == 0 {
+			return nil, BuiltinEmpty{}
 		}
-		var max interface{}
-		for i := range x {
-			if util.Compare(max, x[i]) <= 0 {
-				max = x[i]
+		var max ast.Value = ast.Null{}
+		for i := range a {
+			if ast.Compare(max, a[i].Value) <= 0 {
+				max = a[i].Value
 			}
 		}
-		return ast.InterfaceToValue(max)
+		return max, nil
+	case *ast.Set:
+		if len(*a) == 0 {
+			return nil, BuiltinEmpty{}
+		}
+		max, err := a.Reduce(ast.NullTerm(), func(max *ast.Term, elem *ast.Term) (*ast.Term, error) {
+			if ast.Compare(max, elem) <= 0 {
+				return elem, nil
+			}
+			return max, nil
+		})
+		return max.Value, err
 	}
-	return nil, fmt.Errorf("max: source must be array")
+
+	return nil, builtins.NewOperandTypeErr(1, a, ast.SetTypeName, ast.ArrayTypeName)
+}
+
+func init() {
+	RegisterFunctionalBuiltin1(ast.Count.Name, builtinCount)
+	RegisterFunctionalBuiltin1(ast.Sum.Name, builtinSum)
+	RegisterFunctionalBuiltin1(ast.Max.Name, builtinMax)
```

Comment:
Hrmmm. I wonder if we can use reflection so that:

1. we only need one registration function, e.g.:

```
func RegisterFunctionalBuiltin(fn interface{})
```

2. the built-ins can define their input parameters with concrete types (instead of requiring ast.Value)...for example:

```
def plus(a ast.Number, b ast.Number) (ast.Number, error) {
 ...
}
```